Copter: fix tradheli guided mode takeoff bug

bnsgeyer opened the pull request:
There is a bug in the code that causes helicopters to spool down when a guided mode takeoff is initiated.  I traced the bug to the do_user_takeoff method in Takeoff.cpp.  The call "motors->get_spool_state() was returning something other than throttle unlimited which was causing the code to immediately exit this method and spool down.  I confirmed spool state was throttle unlimited by taking off and landing in loiter prior to initiating the guided takeoff command.

I looked at another instance of this call for spool state and saw it in this form copter.motors->get_spool_state().  I inserted that form into this method and it now seems to function properly.  

I did a grep to see other similar calls for get_spool_state.  I am confused as to what form of this call is to be used. 
![image](https://user-images.githubusercontent.com/9907906/91903762-c8a5f380-ec71-11ea-8a83-aa00eca3eb11.png)

 Most of the calls do not have "copter" in front of them.  What's the right way to do this?  I am told that @peterbarker might be the person to ask.  Help!

peterbarker commented:
@bnsgeyer I really hope the patch here doesn't do anything, or we've got a big problem :-)

A bare "motors" in a Mode object is simple a reference to the same thing on the "copter" object.  We did it this way to avoid code churn and to simplify the transition into using the "Mode" objects.  

Here's the reference declaration:
https://github.com/ardupilot/ardupilot/blob/master/ArduCopter/mode.h#L137

And here's where we initialise it when making the Mode objects:

https://github.com/ardupilot/ardupilot/blob/master/ArduCopter/mode.cpp#L20

You could make sure they're the same thing by printing out the memory the pointer is pointing to with something like:
```
gcs().send_text(MAV_SEVERITY_WARNING, "copter.motors=%p motors=%p", copter.motors, motors);
```

If they're not the same - time to do some headscratching.

bnsgeyer commented:
@peterbarker So I did this last night on my autotune branch which is not totally up to date on master but probably as of May.  which should still have all of these changes.  I made the change in this pr and it worked.  

I tried this pr branch just now and it doesn't work and using your gcs print line displayed that both copter.motors and motors are pointing to the same place.  so now how do I figure out why it is reporting that the spool state is not throttle unlimited?  No I have to do some head scratching. :-(

bnsgeyer commented:
I have updated the PR to address the real issue which was not what I mentioned in the original PR write up. Auto armed is what prevents heli from spooling up. Once I fixed that then another issue came up. Heli’s take much longer to achieve throttle unlimited (spool up to flight rotor speed) than multi’s. So the code was designed to immediately go into takeoff even if the aircraft was at ground idle. Again not an issue for multi’s.  Big issue for heli’s as the wpnav controller is marching ahead with moving the altitude and the heli hasn’t even achieved flight rotor speed. So I added a pretakeoff state to guided mode that pauses the takeoff until the aircraft achieves throttle unlimited state.   
So for heli’s, if they select guided mode before arming then they can turn on motor interlock and nothing will happen until the takeoff command is given. If the aircraft is armed and motor interlock enabled (rotors turning) and then guided mode is selected, then the aircraft will spool down until takeoff command is given. Ultimately, I think a landed state is needed in the guided mode controller.

tridge commented on ArduCopter/system.cpp:
```diff
@@ -416,9 +416,9 @@ void Copter::update_auto_armed()
         // if helicopters are on the ground, and the motor is switched off, auto-armed should be false
         // so that rotor runup is checked again before attempting to take-off
-        if(ap.land_complete && motors->get_spool_state() != AP_Motors::SpoolState::THROTTLE_UNLIMITED && ap.using_interlock) {
+/*        if(ap.land_complete && motors->get_spool_state() != AP_Motors::SpoolState::THROTTLE_UNLIMITED && ap.using_interlock) {
```
we normally don't leave in commented code

peterbarker commented on ArduCopter/mode_guided.cpp:
```diff
@@ -51,39 +51,47 @@ bool ModeGuided::init(bool ignore_checks)
+        guided_mode = Guided_PreTakeOff;
+        _takeoff_alt_cm = takeoff_alt_cm;
+        return true;
+    } else {
```
Don't need this `else`.  Might simplify this delta a lot!

bnsgeyer replied:
why don't I need the else?  it keeps the wpnav controller from being initialized

peterbarker commented:
On Mon, 7 Sep 2020, Bill Geyer wrote:

> +    if (motors->get_spool_state() != AP_Motors::SpoolState::THROTTLE_UNLIMITED) {
> +        guided_mode = Guided_PreTakeOff;
> +        _takeoff_alt_cm = takeoff_alt_cm;
> +        return true;
> +    } else {
> 
> why don't I need the else? it keeps the wpnav controller from being initialized

the "return true" immediately before the else makes the else not needed. 
Either the if condition is true - and you will do the early-return, or it 
is false and you'll currently execute the else block - but if you get rid 
of the else then it will just execute it anyway.

rywunder commented on ArduCopter/mode_guided.cpp:
```diff
@@ -51,39 +51,47 @@ bool ModeGuided::init(bool ignore_checks)
+                takeoff_alt_cm < copter.rangefinder.max_distance_cm_orient(ROTATION_PITCH_270)) {
+            // can't takeoff downwards
+            if (takeoff_alt_cm <= copter.rangefinder_state.alt_cm) {
+                return false;
```
I think you need to add `copter.set_auto_armed(false);` before returning false here so the heli actually spools down.

rywunder commented on ArduCopter/mode_guided.cpp:
```diff
@@ -51,39 +51,47 @@ bool ModeGuided::init(bool ignore_checks)
+            // failure to set destination can only be because of missing terrain data
+            AP::logger().Write_Error(LogErrorSubsystem::NAVIGATION, LogErrorCode::FAILED_TO_SET_DESTINATION);
+            // failure is propagated to GCS with NAK
+            return false;
```
add copter.set_auto_armed(false);

rywunder commented on ArduCopter/mode_guided.cpp:
```diff
@@ -384,6 +396,28 @@ void ModeGuided::run()
+// guided_pretakeoff_run - pretakeoff in guided mode
+//      called by guided_run at 100hz or more
+void ModeGuided::pretakeoff_run()
+{
```
```suggestion
{    // if not armed set throttle to zero and exit immediately
    if (!motors->armed() || !copter.ap.auto_armed) {
        make_safe_spool_down();
        return;
    }
```
We should add this check. Without the check there is undesirable behavior if the user disarms in pretakeoff state. Reproducible in sitl: takeoff command -> disarm while spooling -> rearm and watch the heli spool up without a takeoff command

rmackay9 commented:
@bnsgeyer, I had a look at this and instead of creating a new submode within Guided, I think maybe we can add the same code into the auto_takeoff_run() function.  I've had a try and doing this change in this branch: https://github.com/rmackay9/rmackay9-ardupilot/commits/copter-rough-takeoff-fix

It doesn't make much difference in Copter SITL but the before vs after is shown below.
![copter-smoother-takeoff](https://user-images.githubusercontent.com/1498098/94681289-23f4ff80-035e-11eb-874f-d70bd7b38bb8.png)

rmackay9 commented:
@bnsgeyer I've created PR https://github.com/ArduPilot/ardupilot/pull/15541 with the alternative fix.  I wonder if you could give it a try?  I think the alternative is slightly better because the resulting code is simpler.

bnsgeyer commented:
PR #15541 is the better fix for this bug.  Therefore closing this PR
